### PR TITLE
perf(sampling): batch paged prefill sampler calls

### PR DESCRIPTION
**Pull request by CAICAIIs**
## Summary

Batch paged prefill sampling when no request asks for sample logprobs.

Before this change, `sample_prefill_tokens()` sampled each prefill request independently. For non-greedy requests, that repeated `mx.eval()` + `mlx_to_torch()` + vLLM sampler dispatch once per prefill row.

This PR gathers the last-logit row for each prefill request, stacks them into one `(num_prefill, vocab)` tensor, and calls the existing `SamplingBatch` / vLLM sampler once.

The logprobs path keeps the previous per-request fallback to avoid changing logprob row semantics.

## Scope

In:
- paged prefill sampling
- no-logprobs fast path
- existing `SamplingBatch` and vLLM sampler semantics

Out:
- native MLX non-greedy sampling
- logprobs-path vectorization
- non-paged path changes

Related to #343, but intentionally narrower: this removes repeated sampler bridge work before any native MLX sampler rewrite.

## Performance

Local microbenchmark on Apple M5 / 16GB, vLLM 0.21.0, MLX 0.31.2, vocab=151,936, `temperature=0.8`, `top_p=0.9`:

| batch | old loop | batched prefill sampling | speedup |
| ---: | ---: | ---: | ---: |
| 1 | 2.766 ms | 1.461 ms | 1.89x |
| 4 | 5.881 ms | 2.738 ms | 2.15x |
| 8 | 11.653 ms | 4.692 ms | 2.48x |
| 16 | 23.270 ms | 8.897 ms | 2.62x |

This is a sampling-overhead improvement, not an end-to-end throughput claim.

## Validation

- `python -m pytest tests/test_sampling.py tests/test_v1_model_runner_generate.py -q`
  - `70 passed`
- `python -m ruff check vllm_metal/v1/sampling_batch.py tests/test_sampling.py`
- `python -m ruff format --check vllm_metal/v1/sampling_batch.py tests/test_sampling.py`
- `mypy vllm_metal`
- Real e2e smoke:
  - model: `Qwen/Qwen3-0.6B`
  - Metal paged attention enabled
  - 4 prompt batch
  - `temperature=0.8`, `top_p=0.9`, `max_tokens=8`
  - result: `paged-prefill-sampling-e2e-ok Qwen/Qwen3-0.6B 4`

### CODE REVIEW

**Comment by CAICAIIs** [comment]
related to: https://github.com/vllm-project/vllm-metal/issues/389

**Review by WindChimeRan — COMMENTED** [review]
No end2end performance gain. Can you check?

### Benchmark — long prefill (sonnet 4096+10, 20 prompts, concurrency 32, frac 0.3)

It's using implicit default sampling temperature 0.6. You may explicitly override it. 

<details>
<summary>How to run</summary>

```bash
# server (paged path)
VLLM_METAL_USE_PAGED_ATTENTION=1 VLLM_METAL_MEMORY_FRACTION=0.3 \
  vllm serve Qwen/Qwen3-0.6B --max-model-len 5120

# client
vllm bench serve --backend vllm --base-url http://127.0.0.1:8000 \
  --model Qwen/Qwen3-0.6B --endpoint /v1/completions \
  --dataset-name sonnet --dataset-path sonnet.txt \
  --sonnet-input-len 4096 --sonnet-output-len 10 \
  --num-prompts 20 --request-rate 10 --max-concurrency 32 \
  --ignore-eos --seed 0
```
Qwen3-0.6B · sonnet 4096+10 · 20 prompts, conc 32, frac 0.3 · M1 Pro 32 GB.
`--max-model-len 5120` since the default 2048 can't hold a 4k prompt.

</details>

| Metric | main | PR #390 | Change |
|---|---:|---:|---:|
| Benchmark duration (s) | 71.94 | 75.46 | +4.9% |
| Output tok/s | 2.78 | 2.65 | −4.7% |
| Total tok/s | 1134.0 | 1081.0 | −4.7% |
| Mean TTFT (ms) | 37722.6 | 41131.4 | +9.0% |
| Mean TPOT (ms) | 1566.0 | 1570.4 | +0.3% |
| P99 TPOT (ms) | 1866.5 | 1895.8 | +1.6% |
| Mean ITL (ms) | 1566.0 | 1570.4 | +0.3% |

File: vllm_metal/v1/sampling_batch.py [hunk 1]
```diff
@@ -445,30 +464,50 @@ def sample_prefill_tokens(
...
+        batch = SamplingBatch(
+            sampling_params_list,
+            prompt_token_ids_list,
+            output_token_ids_list,
+            vocab_size=vocab_size,
+            device=device,
+            logitsprocs=logitsprocs,
+            generators=generators,
+        )
+        return sample_from_logits(mx.stack(last_logits_rows), batch, sampler, device)
```
Comment (WindChimeRan) [inline vllm_metal/v1/sampling_batch.py:497]: how about, 

from 
`mx.stack(last_logits_rows)` 

to
`
logits[0, mx.array(last_indices), :]
`
?

(simplify the code to `sample_decode_tokens` pattern)

File: vllm_metal/v1/sampling_batch.py [hunk 2]
```diff
@@ -445,30 +464,50 @@ def sample_prefill_tokens(
     Returns:
         Sampled token IDs and optional logprobs, one row per prefill request.
     """
+    if not prefill_reqs:
```
Comment (WindChimeRan) [inline vllm_metal/v1/sampling_batch.py:467]: possible code simplification:

```python
if not prefill_reqs:
    return _SamplingResult([])

last_indices = [cu_seqlens[num_decode + j + 1] - 1 for j in range(len(prefill_reqs))]
sampling_params_list = [pr.sampling_params for pr in prefill_reqs]
prompt_ids, output_ids = zip(*(_prefill_prompt_and_output_tokens(pr) for pr in prefill_reqs))
generators = {j: pr.generator for j, pr in enumerate(prefill_reqs) if pr.generator is not None}

batch = SamplingBatch(
    sampling_params_list, list(prompt_ids), list(output_ids),
    vocab_size=vocab_size, device=device, logitsprocs=logitsprocs, generators=generators,
)
return sample_from_logits(logits[0, mx.array(last_indices), :], batch, sampler, device)
```

**Comment by CAICAIIs** [comment]
> No end2end performance gain. Can you check?
> 
> ### Benchmark — long prefill (sonnet 4096+10, 20 prompts, concurrency 32, frac 0.3)
> It's using implicit default sampling temperature 0.6. You may explicitly override it.
> 
> How to run
> ```shell
> # server (paged path)
> VLLM_METAL_USE_PAGED_ATTENTION=1 VLLM_METAL_MEMORY_FRACTION=0.3 \
>   vllm serve Qwen/Qwen3-0.6B --max-model-len 5120
> 
> # client
> vllm bench serve --backend vllm --base-url http://127.0.0.1:8000 \
>   --model Qwen/Qwen3-0.6B --endpoint /v1/completions \
>   --dataset-name sonnet --dataset-path sonnet.txt \
>   --sonnet-input-len 4096 --sonnet-output-len 10 \
>   --num-prompts 20 --request-rate 10 --max-concurrency 32 \
>   --ignore-eos --seed 0
> ```
> 
> Qwen3-0.6B · sonnet 4096+10 · 20 prompts, conc 32, frac 0.3 · M1 Pro 32 GB. `--max-model-len 5120` since the default 2048 can't hold a 4k prompt.
> 
> Metric	main	PR #390	Change
> Benchmark duration (s)	71.94	75.46	+4.9%
> Output tok/s	2.78	2.65	−4.7%
> Total tok/s	1134.0	1081.0	−4.7%
> Mean TTFT (ms)	37722.6	41131.4	+9.0%
> Mean TPOT (ms)	1566.0	1570.4	+0.3%
> P99 TPOT (ms)	1866.5	1895.8	+1.6%
> Mean ITL (ms)	1566.0	1570.4	+0.3%

Thanks for the benchmark and sorry for the misjudgment here.

I treated the isolated sampling microbenchmark as enough signal for opening a perf PR, but I should have run a main-vs-PR end-to-end benchmark on a representative serving workload before framing this as `perf`. After reproducing your long-prefill workload locally, I agree this does not show a stable end-to-end gain. The improvement is limited to reducing repeated sampler bridge calls in the no-logprobs paged prefill path, and that overhead is not large enough to move this e2e benchmark reliably.

I also applied your suggested simplification locally by replacing the row list + `mx.stack(...)` path with direct indexed gather:

```python
logits[0, mx.array(last_indices), :]
```
Would you be open to keeping this PR as a small cleanup / sampler-overhead reduction PR, with the title/body changed to avoid any end-to-end performance claim? If that is not useful for the project, I am also happy to close this PR and revisit a larger perf change with clear e2e evidence first. Either way, I learned a lot from your benchmark and review comments here. Thanks again for the guidance.
